---
title: "*Dolosigranulum pigrum* Supplemental Methods"
output: rmarkdown::github_document
---

# SETUP

```{r message=FALSE, include=FALSE}
library(tidyverse)
library(knitr)
library(readr)
knitr::opts_chunk$set(eval = FALSE, message = FALSE)
```

-   [Anvi'o v7, "hope"](https://github.com/merenlab/anvio/releases/tag/v7) was [installed](https://merenlab.org/2016/06/26/installation-v2/) in a Python environment called `anvio-7`.

-   [PPanGGOLiN v1.1.141](https://github.com/labgem/PPanGGOLiN/releases) was [installed](https://github.com/labgem/PPanGGOLiN/wiki/Installation) in a Python environment called `PPanGGOLiN.`

# ANVI'O PANGENOME ANALYSIS

## Reformating genome fasta files

The NCBI submitted .gb genome files were converted to .fa using SnapGene and renamed as indicated on `GENOMES/GenomeNames.txt`. The renamed genome files are available in the `GENOMES/renamed` folder.

These .fa genomes files were reformatted to be compatible with Anvi'o v7. Flag `--seq-type NT` was added to avoid a Config Error indicating the sequence "contains characters that are not any of A, C, T, G, N, a, c, t, g, n."

```{bash}
#conda activate anvio-7
mkdir -p "analysis_Anvio7/Reformat_Dpig"
path_f="GENOMES/renamed"
path_o="analysis_Anvio7/Reformat_Dpig"

for file in $path_f/*.f*; do
    FILENAME=`basename ${file%.*}`
    anvi-script-reformat-fasta -o $path_o/$FILENAME.fa --min-len 0 --simplify-names $file --seq-type NT; 
done
```

## Importing Prokka annotations into Anvi'o

The Prokka annotation was performed with [Prokka v1.14.6](https://github.com/tseemann/prokka/releases/tag/v1.14.6) with default parameters. Prokka annotations were imported into Anvi'o following the recommended [Anvi'o pipeline for Prokka annotated genomes](https://merenlab.org/2017/05/18/working-with-prokka/) described in the next steps.

### Parsing .gff files

The Prokka annotated .gff files are parsed into two tab-delimited text files; one containing the gene calls data and another with the functional annotations. This is done with the script `gff_parser.py`.

```{bash}
#conda activate anvio-7
mkdir -p "analysis_Anvio7/Parsed_prokka_Dpig"
path_f="GENOMES/Prokka_Out_Renamed"
path_o="analysis_Anvio7/Parsed_prokka_Dpig"

for file in $path_f/*.gff; do
    FILENAME=`basename ${file%.*}`
    python analysis_Anvio7/gff_parser.py --gene-calls $path_o/calls_$FILENAME.txt --annotation $path_o/annot_$FILENAME.txt $file;
done
```

### Generating contigs databases

In this step the gene calls from Prokka and the reformatted fasta files get imported to generate Anvi'o contigs databases. We add the `--ignore-internal-stop-codons` flag to avoid errors downstream the pipeline. Only 1-2 genes with internal stop codons were identified in each genome.

```{bash}
#conda activate anvio-7
mkdir -p "analysis_Anvio7/Contigs_db_prokka_Dpig"
path_f="analysis_Anvio7/Reformat_Dpig"
path_o="analysis_Anvio7/Contigs_db_prokka_Dpig"
path_e="analysis_Anvio7/Parsed_prokka_Dpig"

for file in $path_f/*.fa; do
    FILENAME=`basename ${file%.*}`
    anvi-gen-contigs-database -f $file \
                              -o $path_o/$FILENAME.db \
                              --external-gene-calls $path_e/calls_prokka_$FILENAME.txt \
                              --ignore-internal-stop-codons \
                              -n $FILENAME;
done
```

### Prokka annotation

The functional annotation part of the parsed Prokka files gets added to the Anvi'o contigs databases (.db files)

```{bash}
#conda activate anvio-7
path_f="analysis_Anvio7/Contigs_db_prokka_Dpig"
path_e="analysis_Anvio7/Parsed_prokka_Dpig"

for file in $path_f/*.db; do
    FILENAME=`basename ${file%.*}`
    anvi-import-functions -c $file \
                          -i $path_e/annot_prokka_$FILENAME.txt
      
done
```

## Importing other functional annotations

### COG annotation

The [`anvi-run-ncbi-cogs`](https://merenlab.org/software/anvio/help/main/programs/anvi-run-ncbi-cogs/) command was used to annotate the .db genome files against the NCBI's COGs database 2020 release. We used the DIAMOND "sensitive" option instead of the default "fast" one.

```{bash}
#conda activate anvio-7
path_f="analysis_Anvio7/Contigs_db_prokka_Dpig"

for file in $path_f/*.db; do
    anvi-run-ncbi-cogs -T 4 --sensitive -c $file;
done
```

### KEGG annotation

The program [`anvi-run-kegg-kofams`](https://merenlab.org/software/anvio/help/main/programs/anvi-run-kegg-kofams/) annotates the .db genomes with KEGG Orthology (KO) numbers (via hits to the KEGG KOfam database).

```{bash}
#conda activate anvio-7
path_f="analysis_Anvio7/Contigs_db_prokka_Dpig"

for file in $path_f/*.db; do
    anvi-run-kegg-kofams -T 4 -c $file;
done
```

### PFAM annotation

The program [`anvi-run-pfams`](https://merenlab.org/software/anvio/help/main/programs/anvi-run-pfams/) annotates the .db genomes with the PFAM database.

```{bash}
#conda activate anvio-7
path_f="analysis_Anvio7/Contigs_db_prokka_Dpig"

for file in $path_f/*.db; do
    anvi-run-pfams -T 4 -c $file;
done
```

### Running HMMs

The program [`anvi-run-hmms`](https://merenlab.org/software/anvio/help/main/programs/anvi-run-hmms/) stores hmm-hits into the .db genomes files.

```{bash}
#conda activate anvio-7
path_f="analysis_Anvio7/Contigs_db_prokka_Dpig"

for file in $path_f/*.db; do
    anvi-run-hmms -T 4 -c $file;
done
```

### Displaying functions

The program [`anvi-db-info`](https://merenlab.org/software/anvio/help/main/programs/anvi-db-info/) displays information about an Anviâ€™o database. The output of this command in `analysis_Anvio7/AnvioGenomesSummaryFuntions.txt` includes details on all the functional annotations included in the .db files.
```{bash}
#conda activate anvio-7
path_f="analysis_Anvio7/Contigs_db_prokka_Dpig"

for file in $path_f/*.db; do
    anvi-db-info $file;
done
```

## Generating an Anvi'o genomes storage

The program `anvi-gen-genomes-storage` requires you to provide names and locations of the genomes to be included in a genome storage in a .txt file. The  `analysis_Anvio7/AnvioGenomesSummary003.txt` file includes the file names and file paths.

```{bash}
#conda activate anvio-7
anvi-gen-genomes-storage -e analysis_Anvio7/Contigs_db_prokka_Dpig/AnvioGenomesSummary003.txt -o analysis_Anvio7/Pangenomic_Results_Dpig/Anvio005-GENOMES.db --gene-caller Prodigal
```

## Running the pangenomic analysis

We use the program `anvi-pan-genome` to run the actual pangenomic analysis. We used the more sensitive blastp search (`--use-ncbi-blast`) instead of the default DIAMOND fast option. `--mcl-inflation` was set to 10, as recommended for very closely related genomes (i.e., 'strains' of the same 'species').

```{bash}
#conda activate anvio-7
anvi-pan-genome -g analysis_Anvio7/Pangenomic_Results_Dpig/Anvio005-GENOMES.db -o analysis_Anvio7/Pangenomic_Results_Dpig/ -n PAN_DPIG_prokka --use-ncbi-blast --mcl-inflation 10  --num-threads 4
```

## Displaying the pangenome

This will display the pangenome in your default internet browser. Once it opens in the browser, you can adjust image settings and click the "DRAW" icon to view it. For help using the interactive interface see <http://merenlab.org/2016/02/27/the-anvio-interactive-interface/>

```{bash}
#conda activate anvio-7
sudo anvi-display-pan -p analysis_Anvio7/Pangenomic_Results_Dpig/PAN_DPIG_prokka-PAN.db -g analysis_Anvio7/Pangenomic_Results_Dpig/Anvio005-GENOMES.db
```

## Creating bin collections (Core vs Accessory)

In order to define Core vs Accessory pangenome we use Anvi'o interactive interface. In the "Bins" tab we can create and name as many bins as we want and store then in a bin collection. In the "Search" tab we used "Search gene clusters using filters" and "Append splits to selected bin" to create the following bins:

-   Core: 1298/2905 (`r round(100*(1298/2905),1)`%) gene clusters in all 28 genomes

    -   Min number of genomes gene cluster occurs = 28

-   SC Core: 1111/2905 (`r round(100*(1111/2905),1)`%) gene clusters detected in single copy in all 28 genomes

    -   Min number of genomes gene cluster occurs = 28
    -   Max number of genes from each genome = 1

-   MC Core: 1298-1111=187 (`r round(100*(187/2905),1)`%) detected in multiple copy in all 28 genomes

-   SoftCore: 90/2905 (`r round(100*(90/2905),1)`%) gene clusters detected in 26-27 genomes

    -   Min number of genomes gene cluster occurs = 26
    -   Max number of genomes gene cluster occurs = 27

-   Shell: 805/2905 (`r round(100*(805/2905),1)`%) gene clusters detected in 3-25 genomes

    -   Min number of genomes gene cluster occurs = 3
    -   Max number of genomes gene cluster occurs = 25

-   Cloud: 712/2905 (`r round(100*(712/2905),1)`%) gene clusters detected in only 2 genomes or less

    -   Max number of genomes gene cluster occurs = 2

We stored the 5 bins in a new collection named **CorevsAccessory**

We can see the collections and bins created in a pangenomic database using the command `anvi-show-collections-and-bins`

```{bash}
#conda activate anvio-7
anvi-show-collections-and-bins --debug -p analysis_Anvio7/Pangenomic_Results_Dpig/PAN_DPIG_prokka-PAN.db
```

## Summarizing the pangenome

We summarized the pangenome using the program `anvi-summarize`:

```{bash}
#conda activate anvio-7
anvi-summarize -p analysis_Anvio7/Pangenomic_Results_Dpig/PAN_DPIG_prokka-PAN.db \
               -g analysis_Anvio7/Pangenomic_Results_Dpig/Anvio005-GENOMES.db \
               -C CorevsAccessory \
               -o analysis_Anvio7/Pangenomic_Results_Dpig/Dpig-PAN-SUMMARY
```

The resulting summary folder contains the file `analysis_Anvio7/Pangenomic_Results_Dpig/Dpig-PAN-SUMMARY/PAN_DPIG_prokka_gene_clusters_summary.txt.gz` that links each gene to gene clusters, genomes, functions, and bins selected from the interface.

# PPanGGOLiN-Anvi'o 7 ANALYSIS

## Importing Prokka annotations and Anvi'o 7 clusters

In order to import the Anvi'o clustering into PPanGGOLiN we needed:

1.  A .tsv file listing in the first column the gene family names, and in the second column the gene ID that is used in the annotation files.
2.  The annotated genomes with gene IDs that match the ones listed in the previous .tsv file.

Using the `anvi-summarize` output `analysis_Anvio7/Pangenomic_Results_Dpig/Dpig-PAN-SUMMARY/PAN_DPIG_prokka_gene_clusters_summary.txt.gz` described before we can create the .tsv file:

```{r}
Dpig_Anvio7 <- read_delim("analysis_Anvio7/Pangenomic_Results_Dpig/Dpig-PAN-SUMMARY/PAN_DPIG_prokka_gene_clusters_summary.txt.gz", "\t")
Dpig_Anvio7 <- Dpig_Anvio7 %>% 
  unite(new_id, genome_name:gene_callers_id, sep = "___", remove = FALSE)
Clusters_Dpig_Anvio7 <- select(Dpig_Anvio7, c(gene_cluster_id, new_id))
```

```{r}
write_delim(Clusters_Dpig_Anvio7, "analysis_PPanGGOLiN_Anvio7/Clusters_Dpig_Anvio7.tsv", col_names=FALSE)
```

**IMPORTANT:** The gene_callers_id provided by Anvi'o don't match the original ones in the Prokka annotation. The Prokka annotated genomes were parsed into two text files, one for gene calls and one for annotations, with the script `gff_parser.py`. By default Prokka annotates also tRNAs, rRNAs and CRISPR regions. However, `gff_parser.py` will only utilize open reading frames reported by Prodigal in the Prokka output in order to be compatible with the pangenomic Anvi'o pipeline. While parsing new gene_callers_id are generated only for the ORFs that will be imported into Anvi'o. Fortunately `anvi-get-sequences-for-gene-calls` can be used to export new .gff files with only the ORFs and these can be used for PPanGGOLiN, instead of the original Prokka .gff files:

```{bash}
#conda activate anvio-7
mkdir -p "analysis_Anvio7/Exported_gffs"

path_f="analysis_Anvio7/Contigs_db_prokka_Dpig"
path_o="analysis_Anvio7/Exported_gffs"

for file in $path_f/*.db; do
    FILENAME=`basename ${file%.*}`
    anvi-get-sequences-for-gene-calls -c $file --export-gff3 \
                                      -o $path_o/Anvio7_$FILENAME.gff
      
done
```

We created lists with the file names for both the Anvi'o exported .gff (`analysis_PPanGGOLiN_Anvio7/Anvio7GenomesExported.gff.txt`) and the renamed .fasta files (`analysis_PPanGGOLiN_Anvio7/Anvio7GenomesReformatted.fa.txt`) and use them to run the `annotate` PPanGGOLiN subcommand. Then, the `cluster` PPanGGOLiN subcommand was run using the .tsv file created before from the `anvi-summarize` output.

```{bash}
#conda activate PPanGGOLiN
ppanggolin annotate --anno analysis_PPanGGOLiN_Anvio7/Anvio7GenomesExported.gff.txt --fasta analysis_PPanGGOLiN_Anvio7/Anvio7GenomesReformatted.fa.txt -o analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7 --basename FromAnvio7

ppanggolin cluster -p analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/FromAnvio7.h5 --clusters analysis_PPanGGOLiN_Anvio7/Clusters_Dpig_Anvio7.tsv --infer_singletons
```

## Graphing and Partitioning

The PPanGGOLiN`graph` subcommand has only a single other option, which is '-r' or '--remove_high_copy_number'. If used, it will remove the gene families that are too duplicated in your genomes. This is useful if you want to visualize your pangenome afterward and want to remove the biggest hubs to have a clearer view. It can also be used to limit the influence of very duplicated genes such as transposase or ABC transporters in the partition step.

We ran the PPanGGOLiN `graph` and `partition` subcommands with default parameters.

```{bash}
#conda activate PPanGGOLiN
ppanggolin graph -p analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/FromAnvio7.h5

ppanggolin partition -p analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/FromAnvio7.h5
```

## Writing outpus

For details on PPanGGOLiN outputs see: <https://github.com/labgem/PPanGGOLiN/wiki/Outputs>

```{bash}
#conda activate PPanGGOLiN
ppanggolin write -p analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/FromAnvio7.h5 -o analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7 --light_gexf --gexf  --csv --Rtab --stats --partitions --projection --families_tsv -f
```

## Finding Regions of Genome Plasticity

For details on RGPs see: <https://github.com/labgem/PPanGGOLiN/wiki/Regions-of-Genome-Plasticity>

```{bash}
#conda activate PPanGGOLiN
ppanggolin rgp -p analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/FromAnvio7.h5
ppanggolin spot -p analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/FromAnvio7.h5 --draw_hotspots -o analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/spots

ppanggolin write -p analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/FromAnvio7.h5 -o analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7 --regions --spots -f
```

## Summary of used parameters

The PPanGGOLiN `info` subcommand indicates, for each steps of the analysis, the PPanGGOLiN parameters that were used and the source of the data if appropriate.

```{bash}
#conda activate PPanGGOLiN
ppanggolin info -p analysis_PPanGGOLiN_Anvio7/OutputFromAnvio7/FromAnvio7.h5 --parameters
```

The output for this command was:

-   annotation

    -   read_annotations_from_file : True

-   cluster

    -   read_clustering_from_file : True
    -   infer_singletons : True

-   graph

    -   removed_high_copy_number_families : False

-   partition

    -   beta : 2.5
    -   free_dispersion : False
    -   max_node_degree_for_smoothing : 10
    -   computed_K : True
    -   K : 3

-   RGP

    -   persistent_penalty : 3
    -   variable_gain : 1
    -   min_length : 3000
    -   min_score : 4
    -   dup_margin : 0.05

-   spots

    -   set_size : 3
    -   overlapping_match : 2
    -   exact_match : 1
