---
title: \"Genomic Stability and Genetic Defense Systems in *Dolosigranulum pigrum* a Candidate Beneficial Bacterium from the Human Microbiome\" Supplemental Methods
output: rmarkdown::github_document
bibliography: references/references_MGEs.bib
link-citations: true
---

```{r, message = FALSE, include = FALSE}
library(tidyverse)
library(knitr)
library(readr)
library(matrixStats)
library(seqinr)
library(readxl)
knitr::opts_chunk$set(message = FALSE, warning = FALSE)
```

# SELECTED MGEs FROM THE ANVI'O & PPanGGOLiN ANALYSIS

## Anvi'o 7 Annotation search

We import the output of `anvi-summarize` and select the most relevant variables for this analysis:

```{r message=FALSE}
DpigPangenome <-  read_delim("analysis_Anvio7/Pangenomic_Results_Dpig/Dpig-PAN-SUMMARY/PAN_DPIG_prokka_gene_clusters_summary.txt.gz", "\t", escape_double = FALSE, trim_ws = TRUE)
DpigPangenome <- DpigPangenome %>%
  select(-unique_id, -aa_sequence, -SCG, -functional_homogeneity_index, -geometric_homogeneity_index, -combined_homogeneity_index)
```

We search for individual genes annotated as **Transposases**/**Retrons** in either the Prokka [@Seemann2014], COG20 [@Tatusov1997; @Galperin2021], Pfam [@mistry2020] or KOfam [@kanehisa2015; @kanehisa2000] annotations:

```{r}
Retron_Genes <- DpigPangenome %>%
  filter(str_detect(COG20_FUNCTION, "Retron") | str_detect(`Prokka:Prodigal`, "retron") | str_detect(Pfam, "Retron") | str_detect(KOfam, "retron"))
Intron_Genes <- DpigPangenome %>%
  filter(str_detect(COG20_FUNCTION, "Intron") | str_detect(`Prokka:Prodigal`, "intron") | str_detect(Pfam, "Intron") | str_detect(KOfam, "intron"))
Transposase_Genes <- DpigPangenome %>%
  filter(str_detect(COG20_FUNCTION, "Transposase") | str_detect(`Prokka:Prodigal`, "transposase") | str_detect(Pfam, "Transposase")| str_detect(KOfam, "transposase"))
```

There are `r nrow(Retron_Genes)` individual genes annotated as **Retrons**, `r nrow(Intron_Genes)` as **Introns** and `r nrow(Transposase_Genes)` as **Transposases**.

We identified the GCs that contain those individual gene annotations:

```{r}
Retron_GCs <- Retron_Genes %>%
  group_by(gene_cluster_id) %>%
  summarize(n = n())
Intron_GCs <- Intron_Genes %>%
  group_by(gene_cluster_id) %>%
  summarize(n = n())
Transposases_GCs <- Transposase_Genes %>%
  group_by(gene_cluster_id) %>%
  summarize(n = n())
```

There are `r nrow(Retron_GCs)` GCs annotated as **Retrons**, `r nrow(Intron_GCs)` as **Introns** and `r nrow(Transposases_GCs)` as **Transposases**.

### Generating Anvi'o fasta files

From Anvi'o we export the protein Sequences for the GCs with individual genes identified as **Introns**/**Retrons** and as **Transposases**:

For the **Introns**/**Retrons** we focused only on the main GC: GC_00000001:

```{bash, eval=FALSE}
#conda activate anvio-7
anvi-get-sequences-for-gene-clusters -g analysis_Anvio7/Pangenomic_Results_Dpig/Anvio005-GENOMES.db \
                                     -p analysis_Anvio7/Pangenomic_Results_Dpig/PAN_DPIG_prokka-PAN.db \
                                     -o "analysis_MGEs/SequencesAnvio7/GC_00000001_Retron.faa" \
                                     --gene-cluster-id GC_00000001
```

For the **Transposases** we first write a file with the GC numbers for the `r nrow(Transposases_GCs)` identified GCs:

```{r, eval=FALSE}
write.csv(Transposases_GCs$gene_cluster_id, "analysis_MGEs/SequencesAnvio7/IDs_Transposases.txt", row.names = FALSE)
```

(File cleaned up on text editor to removed header and "")

```{bash, eval=FALSE}
#conda activate anvio-7
anvi-get-sequences-for-gene-clusters -g analysis_Anvio7/Pangenomic_Results_Dpig/Anvio005-GENOMES.db \
                                     -p analysis_Anvio7/Pangenomic_Results_Dpig/PAN_DPIG_prokka-PAN.db \
                                     -o "analysis_MGEs/SequencesAnvio7/GCsTransposases.faa" \
                                     --gene-cluster-ids-file "analysis_MGEs/SequencesAnvio7/IDs_Transposases.txt"
```

### Selecting representative sequences

For each GC alignments were visually inspected in AliView and full-length representative sequences selected for PFam search. Using the [PFam batch sequence search](http://pfam.xfam.org/search#tabview=tab1)/[HMMER website](https://www.ebi.ac.uk/Tools/hmmer/search/hmmscan) [@mistry2020; @Potter2018] we classified the initial `r nrow(Transposases_GCs)` putative **Tranposase** GSs as:

-   **Real Transposases:** GC_00000003, GC_00000040, GC_00000055, GC_00001693, GC_00002092, GC_00002210, GC_00002310 and GC_00002501.
-   **Integrases (rve domain):** GC_00000028, GC_00000085, GC_00001701, GC_00001775 and GC_00002348.
-   **Other/Partial:** GC_00000008, GC_00001669, GC_00001787, GC_00002105, GC_00002382, GC_00002430, GC_00002460, GC_00002491, GC_00002679 and GC_00002805.

## OCTAPUS/Clinker

The following files were created with the selected sequences and analyzed with the Bacterial Operon Finder for Functional Organization, aka [OCTAPUS](https://github.com/FredHutch/octapus) to identify the gene neighborhoods in which the selected genes were located across all 28 D. pigrum genomes

-   **SelectedAnvio7_Intron.faa**: Representative sequence for the GC_00000001 cluster.
-   **SelectedAnvio7_Real_Transposases.faa**: GCs initially identified with the word Transposase on the annotation search and with complete (80% coverage or more) PFam **Transposase** domains
-   **SelectedAnvio7_Integrases_rve.faa**: GCs initially identified with the word Transposase on the annotation search but with complete (80% coverage or more) PFam **rve** domains:

The groups of genes identified with OCTAPUS at minimum percent identity 85% and minimum coverage 80% were visualized using [clinker](https://github.com/gamcil/clinker) [@Gilchrist2021]. The following links display the whole clinker output:

-   **Intron:** [GC_00000001](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Intron/GC_00000001/html/GC_00000001_FWD.html?token=ABXYP4EVFJWYFE2WC6AZZ63ANNNXO).
-   **Real Transposases:** [GC_00000003](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Real_Transposases/GC_00000003/html/GC_00000003_FWD.html?token=ABXYP4BN2H7BCSGETPR5JCLANTQWK), [GC_00000040](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Real_Transposases/GC_00000040/html/GC_00000040_FWD.html?token=ABXYP4ARKX24Y4SXK3K6Z3TANTQSG), [GC_00000055](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Real_Transposases/GC_00000055/html/GC_00000055_FWD.html?token=ABXYP4DVD723DXT72EE7PALANTRCK), [GC_00001693](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Real_Transposases/GC_00001693/html/GC_00001693_FWD.html?token=ABXYP4DWBLOH66HM3BPPQBLANTRCW), [GC_00002092](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Real_Transposases/GC_00002092/html/GC_00002092_FWD.html?token=ABXYP4FPCZNI6GFULB5BXSTANTSSS), [GC_00002210](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Real_Transposases/GC_00002210/html/GC_00002210_FWD.html?token=ABXYP4GROU5R72RPLTFMWJ3ANTS2I), [GC_00002310](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Real_Transposases/GC_00002310/html/GC_00002310_FWD.html?token=ABXYP4AIRKRNFNL5F6YOF6DANTTHS) and GC_00002501.
-   **Integrases (rve domain):** [GC_00000028](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Integrases_rve/GC_00000028/html/GC_00000028_FWD.html?token=ABXYP4HY3YKMVHH54ORKPSLANTTBW), [GC_00000085](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Integrases_rve/GC_00000085/html/GC_00000085_FWD.html?token=ABXYP4GGQLFYWG3GPA45YTLANTTEY), [GC_00001701](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Integrases_rve/GC_00001701/html/GC_00001701_FWD.html?token=ABXYP4GWKU3RQ6TRA5APUWDANTTGK), [GC_00001775](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Integrases_rve/GC_00001775/html/GC_00001775_FWD.html?token=ABXYP4H26JYCBI4WNRWY5FTANTTUC) and [GC_00002348](https://htmlpreview.github.io/?https://raw.githubusercontent.com/KLemonLab/DpiMGE_Manuscript/master/analysis_MGEs/BOFFO/Selected_Integrases_rve/GC_00002348/html/GC_00002348_FWD.html?token=ABXYP4HI4UTIO45SAG7BE43ANTURW).

In order to calculate statistics we read the OCTAPUS outputs:

```{r}
Intron <- read_excel("analysis_MGEs/Intron.xlsx")
Real_Transposases <- read_excel("analysis_MGEs/Real_Transposases.xlsx")
Integrases_rve <- read_excel("analysis_MGEs/Integrases_rve.xlsx")
```

GCs renamed with MGE type:

```{r}
Intron <- Intron %>% mutate(MGE = gsub("GC", "INTRON_GC", gene_name))
Real_Transposases <- Real_Transposases %>% mutate(MGE = gsub("GC", "TRANSPOSASE_GC", gene_name))
Integrases_rve <- Integrases_rve %>% mutate(MGE = gsub("GC", "INTEGRASE_GC", gene_name))

ALL <- rbind(Intron, Real_Transposases, Integrases_rve)
```

Summary matrix table with a Genome in each row and each GC listed in each column. Genomes renamed with full IDs:

```{r, message=FALSE}
MatrixOCTAPUS_ALL <- ALL %>%
  group_by(genome_name, MGE) %>%
  summarize(n = n())
MatrixOCTAPUS_ALL <- spread(MatrixOCTAPUS_ALL, MGE, n)
MatrixOCTAPUS_ALL$genomes <- c('ATCC_51524','KPL1914','KPL1922_CDC39_95','KPL1930_CDC2949_98','KPL1931_CDC4294_98','KPL1932_CDC4420_98','KPL1933_CDC4545_98','KPL1934_CDC4709_98','KPL1937_CDC4199_99','KPL1938_CDC4791_99','KPL1939_CDC4792_99','KPL3033','KPL3043','KPL3050','KPL3052','KPL3065','KPL3069','KPL3070','KPL3077','KPL3084','KPL3086','KPL3090','KPL3246','KPL3250','KPL3256','KPL3264','KPL3274','KPL3911')
MatrixOCTAPUS_ALL <- MatrixOCTAPUS_ALL %>% remove_rownames %>% column_to_rownames(var="genomes")
MatrixOCTAPUS_ALL <- select(MatrixOCTAPUS_ALL, -genome_name)
```

Individual tables by MGE:

```{r}
Matrix_Intron <- MatrixOCTAPUS_ALL %>% select(starts_with("INTRON_GC"))
Matrix_Transposases <- MatrixOCTAPUS_ALL %>% select(starts_with("TRANSPOSASE_GC"))
Matrix_Integrases <- MatrixOCTAPUS_ALL %>% select(starts_with("INTEGRASE_GC"))

Matrix_Transposases$Total <- ifelse(apply(is.na(Matrix_Transposases),1,all),NA,rowSums(Matrix_Transposases,na.rm=TRUE))
Matrix_Integrases$Total <- ifelse(apply(is.na(Matrix_Integrases),1,all),NA,rowSums(Matrix_Integrases,na.rm=TRUE))
```

#### Intron Stats

Totals/Stats by row:

```{r}
GCTotal <- colSums(Matrix_Intron, na.rm=TRUE)
GCMean <- round(colMeans(Matrix_Intron, na.rm=TRUE), 2)
GCMedian <- round(colMedians(as.matrix(Matrix_Intron), na.rm=TRUE), 2)
GCVariance <- round(colVars(as.matrix(Matrix_Intron), na.rm=TRUE), 2) #Sample variance
GCSD <- round(colSds(as.matrix(Matrix_Intron), na.rm=TRUE), 2) #Sample standard deviation
GCMAD <- round(colMads(as.matrix(Matrix_Intron), na.rm=TRUE), 2) #Median absolute deviation
GCMin <- round(colMins(as.matrix(Matrix_Intron), na.rm=TRUE), 2) 
GCMax <- round(colMaxs(as.matrix(Matrix_Intron), na.rm=TRUE), 2) 

Matrix_Intron["Total" ,] <- GCTotal
Matrix_Intron["Mean" ,] <- GCMean
Matrix_Intron["Median" ,] <- GCMedian
Matrix_Intron["Variance" ,] <- GCVariance
Matrix_Intron["SD" ,] <- GCSD
Matrix_Intron["MAD" ,] <- GCMAD
Matrix_Intron["Min" ,] <- GCMin
Matrix_Intron["Max" ,] <- GCMax
```

```{r}
opts <- options(knitr.kable.NA = "")
kable(Matrix_Intron, align = "c")
```

#### Transposases Stats

Totals/Stats by row:

```{r}
GCTotal <- colSums(Matrix_Transposases, na.rm=TRUE)
GCMean <- round(colMeans(Matrix_Transposases, na.rm=TRUE), 2)
GCMedian <- round(colMedians(as.matrix(Matrix_Transposases), na.rm=TRUE), 2)
GCVariance <- round(colVars(as.matrix(Matrix_Transposases), na.rm=TRUE), 2) #Sample variance
GCSD <- round(colSds(as.matrix(Matrix_Transposases), na.rm=TRUE), 2) #Sample standard deviation
GCMAD <- round(colMads(as.matrix(Matrix_Transposases), na.rm=TRUE), 2) #Median absolute deviation
GCMin <- round(colMins(as.matrix(Matrix_Transposases), na.rm=TRUE), 2) 
GCMax <- round(colMaxs(as.matrix(Matrix_Transposases), na.rm=TRUE), 2) 

Matrix_Transposases["Total" ,] <- GCTotal
Matrix_Transposases["Mean" ,] <- GCMean
Matrix_Transposases["Median" ,] <- GCMedian
Matrix_Transposases["Variance" ,] <- GCVariance
Matrix_Transposases["SD" ,] <- GCSD
Matrix_Transposases["MAD" ,] <- GCMAD
Matrix_Transposases["Min" ,] <- GCMin
Matrix_Transposases["Max" ,] <- GCMax
```

```{r}
opts <- options(knitr.kable.NA = "")
kable(Matrix_Transposases, align = "c")
```

#### Integrases Stats

Totals/Stats by row:

```{r}
GCTotal <- colSums(Matrix_Integrases, na.rm=TRUE)
GCMean <- round(colMeans(Matrix_Integrases, na.rm=TRUE), 2)
GCMedian <- round(colMedians(as.matrix(Matrix_Integrases), na.rm=TRUE), 2)
GCVariance <- round(colVars(as.matrix(Matrix_Integrases), na.rm=TRUE), 2) #Sample variance
GCSD <- round(colSds(as.matrix(Matrix_Integrases), na.rm=TRUE), 2) #Sample standard deviation
GCMAD <- round(colMads(as.matrix(Matrix_Integrases), na.rm=TRUE), 2) #Median absolute deviation
GCMin <- round(colMins(as.matrix(Matrix_Integrases), na.rm=TRUE), 2) 
GCMax <- round(colMaxs(as.matrix(Matrix_Integrases), na.rm=TRUE), 2) 

Matrix_Integrases["Total" ,] <- GCTotal
Matrix_Integrases["Mean" ,] <- GCMean
Matrix_Integrases["Median" ,] <- GCMedian
Matrix_Integrases["Variance" ,] <- GCVariance
Matrix_Integrases["SD" ,] <- GCSD
Matrix_Integrases["MAD" ,] <- GCMAD
Matrix_Integrases["Min" ,] <- GCMin
Matrix_Integrases["Max" ,] <- GCMax
```

```{r}
opts <- options(knitr.kable.NA = "")
kable(Matrix_Integrases, align = "c")
```

```{r, eval=FALSE}
write.csv(Matrix_Intron, "analysis_MGEs/Intron_Table.csv", row.names = TRUE, na = "")
write.csv(Matrix_Transposases, "analysis_MGEs/Transposases_Table.csv", row.names = TRUE, na = "")
write.csv(Matrix_Integrases, "analysis_MGEs/Integrases_Table.csv", row.names = TRUE, na = "")
```

# [REFERENCES]{.ul}
