---
title: "*Dolosigranulum pigrum* COG ANALYSIS"
output: rmarkdown::github_document
---

```{r, message = FALSE, include = FALSE}
library(tidyverse)
library(readr)
library(forcats)
library(knitr)
library(rstatix)
library(ggpubr)
library(cowplot)
library(grid)
library(RColorBrewer)
```

## Data Cleaning

We import the output of `anvi-summarize` and select the most relevant variables for the functional analysis:
```{r message=FALSE}
DpigPangenome <- read_csv("Dpi_Prokka_Pan_t28_gene_clusters_summary.csv")
DpigPangenome <- DpigPangenome %>%
  select(unique_id, gene_cluster_id, bin_name, genome_name, num_genomes_gene_cluster_has_hits, num_genes_in_gene_cluster, `Prokka:Prodigal_ACC`, `Prokka:Prodigal`, COG_CATEGORY, COG_FUNCTION, COG_FUNCTION_ACC)
```

New variable "accessory_vs_core" where we define "core" as "MC_core"+"SC_core"+"soft_core" and "accessory" as "shell"+"cloud":
```{r}
DpigPangenome<-DpigPangenome %>%
  mutate(accessory_vs_core=ifelse(bin_name=="MC_core"|bin_name=="SC_core"|bin_name=="soft_core","Core","Accessory"))
```

The number of genes and proportions of accessory vs. core are:
```{r, message=FALSE}
nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% filter(accessory_vs_core =="Accessory") %>% summarise)
nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% filter(accessory_vs_core =="Core") %>% summarise)

100*nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% filter(accessory_vs_core =="Accessory") %>% summarise)/nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% summarise)
100*nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% filter(accessory_vs_core =="Core") %>% summarise)/nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% summarise)
```

We define a new variable `COGs` to use in the plots. This variable is based on `COG_CATEGORY` but with a cleaner definition of unclassified, uninformative or mixed assignments: 
* COG categories "Function Unknown" and "General function predictions only" were considered as "Uninformative".
* If the COG category is a mix (for example G|S|M) it gets labeled like "Ambiguous".
* Also missing values (NA) are labeled as Unclassified".
```{r, message=FALSE}
DpigPangenome$COGs <- DpigPangenome$COG_CATEGORY
DpigPangenome$COGs[DpigPangenome$COGs =="S"]<- "Uninformative"
DpigPangenome$COGs[DpigPangenome$COGs =="R"]<- "Uninformative"
DpigPangenome$COGs[grepl('|', DpigPangenome$COGs,fixed=TRUE)]<-"Ambiguous"
DpigPangenome$COGs[is.na(DpigPangenome$COGs)]<-"Unclassified"
```

### Summary

Summary of GOC annotated genes:
```{r}
Summary <- data.frame(
  "Genes" = c("Total in Dpig Pangenome", 
              "COG Category Uninformative = Function Unknown", 
              "COG Category Uninformative = General function predictions only",
              "COG Category Ambiguous (Mixed COG Category)",
              "COG Category Unclassified (Non-assigned)", 
              "Informative COGs (Total - Uninformative, Ambiguous & Unclassified)"),
  "Count" = c(nrow(DpigPangenome), 
              nrow(DpigPangenome %>% filter(COG_CATEGORY =="S")),
              nrow(DpigPangenome %>% filter(COG_CATEGORY =="R")),
              nrow(DpigPangenome %>% filter(COGs =="Ambiguous")),
              nrow(DpigPangenome %>% filter(COGs =="Unclassified")), 
              nrow(DpigPangenome %>% filter(COGs !="Uninformative" & COGs !="Ambiguous" & COGs !="Unclassified"))
              )
)
Summary$Percentage <- round(100*(Summary$Count/49418),2)
```

```{r}
kable(Summary)
```
60.65% of the gene calls are Informative.

## COG analysis corrected by number of genes in each GC

This analysis was done at the pangenomic gene cluster level (GC). Since many gene clusters had mixed COG category assignments a solution is to assign each individual gene call to their corresponding Genome/AccessoryvsCore/COG grouping weighting their contribution by dividing their count by the number of genes in their GC.

The table "DpigCOGsbyGC" groups the genes by genome; and inside genomes by accessory vs. core status, and nested inside as the unambiguous COG category. But, in this case, instead of counting the elements in each group we calculated the sum of 1/num_genes_in_gene_cluster.

```{r, message=FALSE}
DpigCOGsbyGC <-DpigPangenome %>%
  group_by(genome_name, accessory_vs_core, COGs) %>%
    summarise(num_corrected_genes=sum(1/num_genes_in_gene_cluster))
```

The total sum of all values in the `num_corrected_genes` variable should add up to the number of CGs:
```{r, message=FALSE}
sum(DpigCOGsbyGC$num_corrected_genes)
nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% summarise)
```

Extra column to label the gray scale portion of the plots:
```{r}
DpigCOGsbyGC<-DpigCOGsbyGC %>%
  mutate(Assignment=ifelse(COGs!="Uninformative" & COGs!="Ambiguous" & COGs!="Unclassified", " ", COGs))
```

Renaming and ordering factor levels for variables:
```{r}
DpigCOGsbyGC$accessory_vs_core <- factor(DpigCOGsbyGC$accessory_vs_core, levels =c("Core", "Accessory"))

DpigCOGsbyGC$COGs <- recode_factor(DpigCOGsbyGC$COGs, "Q"="Secondary metabolites biosynthesis, transport, and catabolism","P"="Inorganic ion transport and metabolism","I"="Lipid transport and metabolism","H"="Coenzyme transport and metabolism","G"="Carbohydrate transport and metabolism","F"="Nucleotide transport and metabolism","E"="Amino acid transport and metabolism","C"="Energy production and conversion","X"="Mobilome: prophages, transposons","L"="Replication, recombination and repair","K"="Transcription","J"="Translation, ribosomal structure and biogenesis","V"="Defense mechanisms","U"="Intracellular trafficking, secretion, and vesicular transport","T"="Signal transduction mechanisms","O"="Post-translational modification, protein turnover, and chaperones","N"="Cell Motility","M"="Cell wall/membrane/envelope biogenesis","D"="Cell cycle control, cell division, chromosome partitioning","Uninformative"="Uninformative","Ambiguous"="Ambiguous","Unclassified"="Unclassified", .ordered = TRUE)

DpigCOGsbyGC$Assignment <- factor(DpigCOGsbyGC$Assignment, levels =c(" ", "Uninformative", "Ambiguous", "Unclassified"))

DpigCOGsbyGC$genome_name <- recode_factor(DpigCOGsbyGC$genome_name, "Dpigrum_ATCC_51524"="ATCC 51524", "Dpigrum_KPL3250"="KPL3250", "Dpigrum_KPL1939_CDC4792_99"="CDC 4792-99","Dpigrum_KPL1934_CDC4709_98"="CDC 4709-98", "Dpigrum_KPL1922_CDC39_95"="CDC 39-95", "Dpigrum_KPL3264"="KPL3264", "Dpigrum_KPL3256"="KPL3256", "Dpigrum_KPL3033"="KPL3033", "Dpigrum_KPL1933_CDC4545_98"="CDC 4545-98", "Dpigrum_KPL1930_CDC2949_98"="CDC 2949-98", "Dpigrum_KPL3069"="KPL3069", "Dpigrum_KPL3052"="KPL3052", "Dpigrum_KPL3090"="KPL3090", "Dpigrum_KPL3086"="KPL3086", "Dpigrum_KPL3065"="KPL3065", "Dpigrum_KPL3043"="KPL3043", "Dpigrum_KPL3911"="KPL3911", "Dpigrum_KPL3084"="KPL3084", "Dpigrum_KPL3070"="KPL3070",
"Dpigrum_KPL3246"="KPL3246", "Dpigrum_KPL1937_CDC4199_99"="CDC 4199-99","Dpigrum_KPL3274"="KPL3274","Dpigrum_KPL3050"="KPL3050","Dpigrum_KPL1938_CDC4791_99"="CDC 4791-99", "Dpigrum_KPL1932_CDC4420_98"="CDC 4420-98", "Dpigrum_KPL3077"="KPL3077", "Dpigrum_KPL1931_CDC4294_98"="CDC 4294-98", "Dpigrum_KPL1914"="KPL1914", .ordered = TRUE)
```

Color Palettes
```{r}
getPalette <- colorRampPalette(brewer.pal(8, "Set1"))
CountTotalCOGs <- length(unique(DpigCOGsbyGC$COGs))

palette1 <- c("grey60", "grey40", "grey20", getPalette(CountTotalCOGs-3)) # 22 elements: Colors + Grays
palette2 <- getPalette(CountTotalCOGs-3) # 19 elements: Colors
palette3 <- c("grey60", "grey40", "grey20", "white") # 4 elements: White + Grays
```

### Plots Accessory vs. Core

Panel A in main figure:
```{r, TotalGCs.accessory_vs_core}
pA <- ggplot(DpigCOGsbyGC, aes(x = accessory_vs_core, y = num_corrected_genes, fill = fct_rev(COGs))) +
  stat_summary(fun=sum ,geom="bar", position = "stack") +
  scale_x_discrete(labels = c("Soft/Core", "Accessory")) +
  scale_fill_manual(values = palette1) +
  scale_y_continuous(expand = c(0,0), breaks=seq(0, 1500, by = 250)) +
  labs(fill="COG Categories", x=" ", y= "Number of Gene Clusters") +
  theme_classic() +
  theme(axis.title = element_text(size = 9), axis.text = element_text(size=7), plot.margin=unit(c(10,0,10,20),"pt"), legend.position = "none") 
pA
```

This plot is used for the grayscale legend:
```{r, InformativeGCs.accessory_vs_core}
pB <- ggplot(DpigCOGsbyGC, aes(x = accessory_vs_core, y = num_corrected_genes, fill = fct_rev(Assignment))) +
  stat_summary(fun=sum ,geom="bar", position = "stack") +
  scale_x_discrete(labels = c("Soft/Core", "Accessory")) +
  scale_fill_manual(values = palette3) +
  scale_y_continuous(expand = c(0,0), breaks=seq(0, 1500, by = 250)) +
  labs(fill=" ", x=" ", y= "Number of Gene Clusters") +
  theme_classic() +
  theme(axis.title = element_text(size = 9), axis.text = element_text(size=7), plot.margin=unit(c(10,0,10,20),"pt"), legend.key.size = unit(0.7, "line"), legend.text = element_text(size = 7), legend.box.margin = margin(10,20,10,10)) +
  guides(fill=guide_legend(ncol=1, title.position = "top", title.hjust = 0.5))
pB
```

### Plots by Genome

Panel A in supplemental figure:
```{r, TotalGCs.accesory.byGenome}
pC <- ggplot(filter(DpigCOGsbyGC, accessory_vs_core == "Accessory"), aes(x=genome_name, y=num_corrected_genes, fill = fct_rev(COGs))) +
  stat_summary(fun=sum ,geom="bar", position = "stack") +
  scale_fill_manual(values = palette1) +
  scale_y_continuous(expand = c(0,0)) + 
  labs(fill="COG Assignment", x="", y= "Number of Gene Clusters") +
  theme_classic() + 
  theme(axis.text.y = element_text(size=7), axis.text.x = element_text(size=8, angle=75, hjust=1)) +
  theme(legend.position = "none", plot.margin=unit(c(15,15,-10,20),"pt")) 

pC
```
Panel B in supplemental figure:
```{r, InformativeGCs.accesory.byGenome}
pD <- ggplot(filter(DpigCOGsbyGC %>% filter(COGs != "Uninformative", COGs !="Ambiguous", COGs != "Unclassified"), accessory_vs_core == "Accessory"), aes(x=genome_name, y=num_corrected_genes, fill = fct_rev(COGs))) +
  stat_summary(fun=sum ,geom="bar", position = "stack") +
  scale_y_continuous(expand = c(0,0)) + 
  scale_fill_manual(values = palette2) + 
  labs(fill="COG Categories", x="", y= "Number of Informative Gene Clusters") +
  theme_classic() + 
  theme(axis.text.y = element_text(size=7), axis.text.x = element_text(size=8, angle=75, hjust=1)) +
  theme(legend.position="bottom", legend.key.size = unit(0.7, "line"), legend.text = element_text(size = 8), plot.margin=unit(c(0,15,0,20),"pt")) +
  guides(fill=guide_legend(ncol=2, title.position = "top", title.hjust = 0.5)) 
pD
```
This plot is used for the grayscale legend:
```{r, TotalGCs.accesory.byGenome.legend}
pE <- ggplot(filter(DpigCOGsbyGC, accessory_vs_core == "Accessory"), aes(x=genome_name, y=num_corrected_genes, fill = fct_rev(Assignment))) +
  stat_summary(fun=sum ,geom="bar", position = "stack") +
  scale_fill_manual(values = palette3) +
  scale_y_continuous(expand = c(0,0)) + 
  labs(fill="Accessory Genome COG Assignment", x="", y= "Number of Gene Clusters") +
  theme_classic() + 
  theme(axis.text.y = element_text(size=7), axis.text.x = element_text(size=8, angle=75, hjust=1)) +
  theme(legend.position="bottom", legend.key.size = unit(0.7, "line"), legend.text = element_text(size = 8), legend.title = element_text(face="bold", size = 12), plot.margin=unit(c(15,15,0,20),"pt")) +
  guides(fill=guide_legend(nrow=1, title.position = "top", title.hjust = -5))
pE
```

### Plots by COG Category

The table "COGTotalGC" groups the genes by accessory vs. core status, and nested inside as the unambiguous COG category. Table converted to the wide format.
```{r, message=FALSE}
COGTotalGC<-DpigPangenome %>%
  group_by(accessory_vs_core, COGs) %>%
  summarise(num_corrected_genes=sum(1/num_genes_in_gene_cluster))

COGTotalGC$COGs <- recode_factor(COGTotalGC$COGs, "Q"="Secondary metabolites biosynthesis, transport, and catabolism","P"="Inorganic ion transport and metabolism","I"="Lipid transport and metabolism","H"="Coenzyme transport and metabolism","G"="Carbohydrate transport and metabolism","F"="Nucleotide transport and metabolism","E"="Amino acid transport and metabolism","C"="Energy production and conversion","X"="Mobilome: prophages, transposons","L"="Replication, recombination and repair","K"="Transcription","J"="Translation, ribosomal structure and biogenesis","V"="Defense mechanisms","U"="Intracellular trafficking, secretion, and vesicular transport","T"="Signal transduction mechanisms","O"="Post-translational modification, protein turnover, and chaperones","N"="Cell Motility","M"="Cell wall/membrane/envelope biogenesis","D"="Cell cycle control, cell division, chromosome partitioning","Uninformative"="Uninformative","Ambiguous"="Ambiguous","Unclassified"="Unclassified", .ordered = TRUE)

COGTotalGC <- spread(COGTotalGC, accessory_vs_core, num_corrected_genes)
```

```{r}
A <- round(100*sum(COGTotalGC$Accessory)/(sum(COGTotalGC$Accessory)+sum(COGTotalGC$Core)), 2)
C <- round(100*sum(COGTotalGC$Core)/(sum(COGTotalGC$Accessory)+sum(COGTotalGC$Core)), 2)
```

The proportions are `r A`% accessory vs. `r C`% core at the pangenome level. We calculate the accessory vs. core proportions by COG category, to see the COG categories that are enriched in the accessory. 
```{r}
COGTotalGC$total <- COGTotalGC$Accessory + COGTotalGC$Core
COGTotalGC$p.accessory <- round(100*(COGTotalGC$Accessory/COGTotalGC$total), 2)
COGTotalGC$p.core <- round(100*(COGTotalGC$Core/COGTotalGC$total), 2)
COGTotalGC$enrichment <- round(COGTotalGC$p.accessory/A, 2)
```

```{r}
kable(COGTotalGC)
```

In order to represent the Core on the left of a plot with absolute values per COG category we create `core.neg`; a negative version of the `core` variable in COGTotalGC. Table converted to the long format for plotting.
```{r}
COGTotalGC$core.neg <- -COGTotalGC$Core
COGTotalGCLong <- gather(COGTotalGC, accessory_vs_core, plotting, core.neg, Accessory)
```

```{r, InformativeGCs.byCOG}
pF <- ggplot(filter(COGTotalGCLong, COGs != "Uninformative", COGs != "Ambiguous", COGs != "Unclassified"), aes(x = COGs, y = plotting, fill = COGs)) +
  geom_bar(stat="identity") + 
  scale_fill_manual(values = rev(palette2)) + 
  scale_x_discrete(position = "top") +
  labs(title= "COG Categories", x="", y= "Number of Gene Clusters") +
  coord_flip() +
  scale_y_continuous(limits = c(-170, 170), breaks = c(-150, -100, -50, 0, 50, 100, 150), label = c(150, 100, 50, 0, 50, 100, 150)) +
  geom_segment(aes(x=0,xend=19.5,y=0,yend=0)) +
  geom_label(aes(x = 20.5, y = -95, label = "      Soft/Core       "), fontface="bold", size=3, fill = "grey90", label.size=NA, label.padding = unit(0.3, "lines")) +
  geom_label(aes(x = 20.5, y = 95, label = "     Accessory      "), fontface="bold", size=3, fill = "grey90", label.size=NA, label.padding = unit(0.3, "lines")) +
  theme_classic() +
  theme(axis.title = element_text(size = 9), axis.text.x = element_text(size=7), axis.ticks.y = element_blank(), axis.line.y = element_blank(), legend.position = "none", plot.margin=unit(c(5,10,10,25),"pt"), plot.title=element_text(face="bold", hjust=3, vjust=-3.9)) 

pF
gpF <- ggplotGrob(pF)
gpF$layout$clip[gpF$layout$name=="panel"] <- "off"
```

This is used for the clade labels: 
```{r}
clades <- ggplot() +
  scale_y_continuous(limits = c(-1.5, 0.5), breaks = c(-1, 0)) +
  geom_segment(aes(x=0,xend=2.9,y=0,yend=0), color="#2c9b51ff") +
  geom_segment(aes(x=3,xend=4.9,y=0,yend=0), color="#a851a8ff") +
  geom_segment(aes(x=5,xend=9.9,y=0,yend=0), color="#e17139ff") +
  geom_segment(aes(x=10,xend=28,y=0,yend=0), color="#1b1d86ff") +
  annotate("text", x = 1.5, y = -1, label = "C1", fontface="bold", color="#2c9b51ff")+
  annotate("text", x = 4, y = -1, label = "C2", fontface="bold", color="#a851a8ff")+
  annotate("text", x = 7.3, y = -1, label = "C3", fontface="bold", color="#e17139ff")+
  annotate("text", x = 19, y = -1, label = "C4", fontface="bold", color="#1b1d86ff")+
  theme_classic() +
  theme(axis.title = element_blank(), axis.text = element_blank(), axis.ticks = element_blank(), axis.line = element_blank(), plot.margin=unit(c(0,0,5,20),"pt")) 
clades
```

## Final Figures

```{r}
pMain <- ggarrange(ggarrange(get_legend(pB), pA, ncol = 1, heights = c(0.2, 1)),
                   gpF, ncol = 2, labels = c("A", "B"), hjust=-0.5, vjust=2, widths = c(0.7, 2))

ggsave("Fig4_COG_summary.tiff", pMain, width = 9, height = 4, dpi = 150)
```

```{r}
pSupple <- ggarrange(get_legend(pE),
                      ggarrange(pC+theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()), pD+theme(legend.position="none"), ncol = 1,  align = "v", labels = c("A", "B"), hjust=-0.5, vjust=1, heights = c(1, 1)),
                      clades, 
                      get_legend(pD), ncol = 1, heights = c(0.2, 2, 0.2, 0.6))

ggsave("FigS4_COG_byGenome.tiff", pSupple, width = 8, height = 10, dpi = 150)
```












