---
title: "*Dolosigranulum pigrum* COG ANALYSIS"
output: rmarkdown::github_document
---

```{r, message = FALSE, include = FALSE}
library(tidyverse)
library(readr)
library(knitr)
library(rstatix)
library(ggpubr)
library(RColorBrewer)
```

## Data Input

We import the output of `anvi-summarize` and select the most relevant variables for the functional analysis:
```{r message=FALSE}
DpigPangenome <- read_csv("Dpi_Prokka_Pan_t28_gene_clusters_summary.csv")
DpigPangenome <- DpigPangenome %>%
  select(unique_id, gene_cluster_id, bin_name, genome_name, num_genomes_gene_cluster_has_hits, num_genes_in_gene_cluster, `Prokka:Prodigal_ACC`, `Prokka:Prodigal`, COG_CATEGORY, COG_FUNCTION, COG_FUNCTION_ACC)
```

New variable "accessory_vs_core" where we define "core" as "MC_core"+"SC_core"+"soft_core" and "accessory" as "shell"+"cloud":
```{r}
DpigPangenome<-DpigPangenome %>%
  mutate(accessory_vs_core=ifelse(bin_name=="MC_core"|bin_name=="SC_core"|bin_name=="soft_core","core","accessory"))
```

The number of genes and proportions of accessory vs. core are:
```{r, message=FALSE}
nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% filter(accessory_vs_core =="accessory") %>% summarise)
nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% filter(accessory_vs_core =="core") %>% summarise)

100*nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% filter(accessory_vs_core =="accessory") %>% summarise)/nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% summarise)
100*nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% filter(accessory_vs_core =="core") %>% summarise)/nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% summarise)
```

New variables to identify the ambiguous COG annotations calls, both at the COG category and COG function levels.
* If the COG_FUNCTION_ACC is ambiguous (for example COG4409|COG4932) it gets labeled like a mix function ("mixF"). 
* If the COG_CATEGORY is ambiguous (for example G|S|M) it gets labeled like a mix category ("mixC"). 
* Also missing values are flagged as "NA".
```{r}
DpigPangenome$COG_FUNCTION_CLEAN <- DpigPangenome$COG_FUNCTION_ACC
DpigPangenome$COG_FUNCTION_CLEAN[is.na(DpigPangenome$COG_FUNCTION_ACC)]<-"NA"
DpigPangenome$COG_FUNCTION_CLEAN[grepl('|', DpigPangenome$COG_FUNCTION_ACC,fixed=TRUE)]<-"mixF"

DpigPangenome$COG_CATEGORY_CLEAN <- DpigPangenome$COG_CATEGORY
DpigPangenome$COG_CATEGORY_CLEAN[is.na(DpigPangenome$COG_CATEGORY)]<-"NA"
DpigPangenome$COG_CATEGORY_CLEAN[grepl('|', DpigPangenome$COG_CATEGORY,fixed=TRUE)]<-"mixC"
```

New variable with all the ambiguous calls combined:
```{r}
DpigPangenome$UnambiguousCategory <- DpigPangenome$COG_CATEGORY
DpigPangenome$UnambiguousCategory[is.na(DpigPangenome$COG_FUNCTION_ACC)]<- "Ambiguous"
DpigPangenome$UnambiguousCategory[DpigPangenome$COG_CATEGORY_CLEAN =="S"]<- "Ambiguous"
DpigPangenome$UnambiguousCategory[DpigPangenome$COG_CATEGORY_CLEAN =="R"]<- "Ambiguous"
DpigPangenome$UnambiguousCategory[DpigPangenome$COG_CATEGORY_CLEAN =="mixC"]<- "Ambiguous"
```

## Data Summary

Summary of GOC annotated genes:
```{r}
Summary <- data.frame(
  "Genes" = c("Total in Dpig Pangenome", 
              "COG Category = Non-assigned", 
              "COG Category = Function Unknown", 
              "COG Category = General function predictions only",
              "COG Category = Mixed",
              "Unambiguous"),
  "Count" = c(nrow(DpigPangenome), 
              nrow(DpigPangenome %>% filter(COG_FUNCTION_CLEAN =="NA")), 
              nrow(DpigPangenome %>% filter(COG_CATEGORY_CLEAN =="S")),
              nrow(DpigPangenome %>% filter(COG_CATEGORY_CLEAN =="R")),
              nrow(DpigPangenome %>% filter(COG_CATEGORY_CLEAN =="mixC")),
              nrow(DpigPangenome %>% filter(UnambiguousCategory !="Ambiguous"))
              )
)
Summary$Percentage <- round(100*(Summary$Count/49418),2)
```
```{r}
kable(Summary)
```
60.65% of the gene calls are Unambiguous: this is COG Category is Non-assigned, Function Unknown, General function predictions only or a mixed Category.

## COG analysis corrected by number of genes in each GC

This analysis was done at the pangenomic gene cluster level (GC). Since many gene clusters had mixed COG category assignments a solution is to assign each individual gene call to their corresponding Genome/AccessoryvsCore/COG grouping weighting their contribution by dividing their count by the number of genes in their GC.

The table "DpigCOGsbyGC" groups the genes by genome; and inside genomes by accessory vs. core status, and nested inside as the unambiguous COG category. But, in this case, instead of counting the elements in each group we calculated the sum of 1/num_genes_in_gene_cluster.

```{r, message=FALSE}
DpigCOGsbyGC <-DpigPangenome %>%
  group_by(genome_name, accessory_vs_core, UnambiguousCategory) %>%
    summarise(num_corrected_genes=sum(1/num_genes_in_gene_cluster))
```

The total sum of all values in the `num_corrected_genes` variable should add up to the number of CGs:
```{r, message=FALSE}
sum(DpigCOGsbyGC$num_corrected_genes)
nrow(DpigPangenome %>% group_by(gene_cluster_id) %>% summarise)
```

Defining order of COG categories and Strains
```{r}
DpigCOGsbyGC$accessory_vs_core <- factor(DpigCOGsbyGC$accessory_vs_core)

DpigCOGsbyGC$UnambiguousCategory <- factor(DpigCOGsbyGC$UnambiguousCategory, levels = c("D","M","N","O","T","U","V","J","K","L","X","C","E","F","G","H","I","P","Q","Ambiguous"))
COGlabels = c("Cell cycle control, cell division, chromosome partitioning","Cell wall/membrane/envelope biogenesis","Cell Motility","Post-translational modification, protein turnover, and chaperones","Signal transduction mechanisms","Intracellular trafficking, secretion, and vesicular transport","Defense mechanisms","Translation, ribosomal structure and biogenesis","Transcription","Replication, recombination and repair","Mobilome: prophages, transposons","Energy production and conversion","Amino acid transport and metabolism","Nucleotide transport and metabolism","Carbohydrate transport and metabolism","Coenzyme transport and metabolism","Lipid transport and metabolism","Inorganic ion transport and metabolism","Secondary metabolites biosynthesis, transport, and catabolism","Ambiguous")

DpigCOGsbyGC$genome_name <- factor(DpigCOGsbyGC$genome_name, levels = c("Dpigrum_ATCC_51524", "Dpigrum_KPL1939_CDC4792_99","Dpigrum_KPL3250","Dpigrum_KPL1922_CDC39_95","Dpigrum_KPL1934_CDC4709_98","Dpigrum_KPL1930_CDC2949_98","Dpigrum_KPL1933_CDC4545_98","Dpigrum_KPL3033","Dpigrum_KPL3256","Dpigrum_KPL3264","Dpigrum_KPL1914","Dpigrum_KPL1931_CDC4294_98","Dpigrum_KPL3077","Dpigrum_KPL1932_CDC4420_98","Dpigrum_KPL1938_CDC4791_99","Dpigrum_KPL3050","Dpigrum_KPL3274","Dpigrum_KPL1937_CDC4199_99","Dpigrum_KPL3246","Dpigrum_KPL3070","Dpigrum_KPL3084","Dpigrum_KPL3911","Dpigrum_KPL3043","Dpigrum_KPL3065","Dpigrum_KPL3086","Dpigrum_KPL3090","Dpigrum_KPL3052","Dpigrum_KPL3069"))
strainlabels = c("ATCC 51524", "CDC 4792-99","KPL3250","CDC39-95","CDC 4709-98","CDC 2949-98","CDC 4545-98","KPL3033","KPL3256","KPL3264","KPL1914","CDC 4294-98","KPL3077","CDC 4420-98","CDC 4791-99","KPL3050","KPL3274","CDC 4199-99","KPL3246","KPL3070","KPL3084","KPL3911","KPL3043","KPL3065","KPL3086","KPL3090","KPL3052","KPL3069")
```

Color Palette
```{r}
colourCount = length(unique(DpigCOGsbyGC$UnambiguousCategory))
getPalette = colorRampPalette(brewer.pal(9, "Set1"))
```

### COG Plots Accessory vs. Core

```{r, accessory_vs_core}
ggplot(DpigCOGsbyGC, aes(x = accessory_vs_core, y = num_corrected_genes, fill = UnambiguousCategory)) +
  stat_summary(fun=sum ,geom="bar", position = "stack") +
  scale_fill_manual(values = getPalette(colourCount), labels=COGlabels) +
  scale_y_continuous(expand = c(0,0)) + 
  labs(fill="COG Categories", x="", y= "Number of Genes", title="Total number of genes in the Accesory vs Core (fun=sum)") +
  theme_classic() +
  theme(legend.key.size = unit(0.45, "cm")) 
```

### COG Plots by Genome

```{r, accesory.byCOG.byGenome}
ggplot(filter(DpigCOGsbyGC, accessory_vs_core == "accessory"), aes(x=genome_name, y=num_corrected_genes, fill=UnambiguousCategory)) +
  geom_bar(stat="identity") + 
  scale_y_continuous(expand = c(0,0)) + 
  scale_x_discrete(labels = strainlabels) +
  scale_fill_manual(values = getPalette(colourCount), labels=COGlabels) + 
  labs(fill="COG Categories", x="", y= "Number of Genes", title="Accesory genes by COG category") +
  theme_classic() + 
  theme(axis.text.x = element_text(size=8, angle=75, hjust=1)) +
  theme(legend.position="right",legend.key.size = unit(0.3, "cm")) +
  guides(fill=guide_legend(ncol=1, title.position = "top")) 
```

```{r, accesory.byUnambiguousCOG.byGenome}
ggplot(filter(DpigCOGsbyGC %>% filter(UnambiguousCategory !="Ambiguous"), accessory_vs_core == "accessory"), aes(x=genome_name, y=num_corrected_genes, fill=UnambiguousCategory)) +
  geom_bar(stat="identity") + 
  scale_y_continuous(expand = c(0,0)) + 
  scale_x_discrete(labels = strainlabels) +
  scale_fill_manual(values = getPalette(colourCount), labels=COGlabels) + 
  labs(fill="COG Categories", x="", y= "Number of Genes", title="Unambiguous accesory genes by COG category") +
  theme_classic() + 
  theme(axis.text.x = element_text(size=8, angle=75, hjust=1)) +
  theme(legend.position="right",legend.key.size = unit(0.3, "cm")) +
  guides(fill=guide_legend(ncol=1, title.position = "top")) 
```

### COG Plots by COG Category

```{r, accessory_vs_core.byCOG}
ggplot(DpigCOGsbyGC, aes(x = UnambiguousCategory, y = num_corrected_genes, fill = accessory_vs_core)) +
  stat_summary(fun=sum ,geom="bar", position = "stack") +
  scale_y_continuous(expand = c(0,0)) + 
  labs(fill="", x="COG Categories", y= "Number of GCs", title="GC by COG category") +
  scale_x_discrete(labels = rev(COGlabels), limits = rev(levels(DpigCOGsbyGC$UnambiguousCategory))) +
  theme_classic() +
  coord_flip()
```

The table "COGTotalGC" groups the genes by accessory vs. core status, and nested inside as the unambiguous COG category. Table converted to the wide format.
```{r, message=FALSE}
COGTotalGC<-DpigPangenome %>%
  group_by(accessory_vs_core, UnambiguousCategory) %>%
  summarise(num_corrected_genes=sum(1/num_genes_in_gene_cluster))

COGTotalGC$UnambiguousCategory <- factor(COGTotalGC$UnambiguousCategory, levels = c("D","M","N","O","T","U","V","J","K","L","X","C","E","F","G","H","I","P","Q","Ambiguous"))

COGTotalGC <- spread(COGTotalGC, accessory_vs_core, num_corrected_genes)
```

```{r}
A <- round(100*sum(COGTotalGC$accessory)/(sum(COGTotalGC$accessory)+sum(COGTotalGC$core)), 2)
C <- round(100*sum(COGTotalGC$core)/(sum(COGTotalGC$accessory)+sum(COGTotalGC$core)), 2)
```

The proportions are `r A`% accessory vs. `r C`% core at the pangenome level. We calculate the accessory vs. core proportions by COG category, to see the COG categories that are enriched in the accessory. 
```{r}
COGTotalGC$total <- COGTotalGC$accessory + COGTotalGC$core
COGTotalGC$p.accessory <- round(100*(COGTotalGC$accessory/COGTotalGC$total), 2)
COGTotalGC$p.core <- round(100*(COGTotalGC$core/COGTotalGC$total), 2)
COGTotalGC$enrichment <- round(COGTotalGC$p.accessory/A, 2)
```

```{r}
kable(COGTotalGC)
```

```{r, accesory.byUnambiguousCOG.byGenome}
ggplot(COGTotalGC, aes(x=UnambiguousCategory, y=enrichment)) +
  geom_bar(stat="identity") + 
  scale_y_continuous(trans = "log2", expand = c(0,0)) + 
  labs(x="", y= "log2(% COG in accesory / average % accesory)", title="GC Enrichment by COG category") +
  scale_x_discrete(labels = rev(COGlabels), limits = rev(levels(DpigCOGsbyGC$UnambiguousCategory))) +
  coord_flip() +
  theme_classic() +
  theme(axis.ticks.y = element_blank(), axis.line.y = element_blank()) +
  geom_hline(yintercept=1)
```
